Add docs for `removeTypenameFromVariables` link (#11104)
apollographql/apollo-client: docs/source/api/link/apollo-link-remove-typename.mdx

// {
//   "dashboard": {
//     "__typename": "Dashboard",
//     "id": 1,
//     "name": "My Dashboard"
//   }
// }
```

Now let's update this dashboard by sending a mutation to our server. We'll use the `dashboard` as input to our mutation.

```ts
const mutation = gql`
  mutation UpdateDashboardMutation($dashboard: DashboardInput!) {
    updateDashboard(dashboard: $dashboard) {
      id
      name
    }
  }
`;

await client.mutate({ 
  mutation,
  variables: { 
    dashboard: { ...data.dashboard, name: 'My Updated Dashboard' } 
  }
});
```

Without the use of the `removeTypenameFromVariables` link, the server will return an error because `data.dashboard` still contains the `__typename` field.

## Keep `__typename` in JSON scalars

While the [GraphQL spec](https://spec.graphql.org/October2021/#sec-Input-Objects.Type-Validation) disallows input fields that begin with the characters "\_\_" (two underscores), this restriction is not applied when the input field is a JSON scalar (i.e. a scalar type that accepts raw JSON as input). When your server relies on `__typename` within certain `JSON` scalars, you can configure the `removeTypenameFromVariables` link to retain `__typename`.

> > Note: the JSON scalar type does not need to be literally named `JSON` to be considered a JSON scalar.

Reviewer: I think I'm going to make a shared component so we can standardize how we make these callouts :)

---
title: Remove Typename Link
description: Automatically remove __typename fields from variables.
minVersion: 3.8.0
---

## Overview

> Use the `removeTypenameFromVariables` link to automatically remove `__typename` fields from variables in operations. This is useful when you're reusing data from a query as an argument to another GraphQL operation.

Reviewer 2: ```suggestion
You can use the `removeTypenameFromVariables` link to automatically remove `__typename` fields from variables in operations. This is useful when you're reusing data from a query as an argument to another GraphQL operation.
```

Not sure if we have a general style to follow, but as a reader I'd prefer this to start a little less direct.

Reviewer 3: +1 to Lenz's suggestion, just further tweaking it to stay consistent with `BatchHttpLink` and `HttpLink` opening paragraphs:

```suggestion
The `removeTypenameFromVariables` link automatically removes `__typename` fields from variables in operations. This is useful when reusing data from a query as an argument to another GraphQL operation.
```

Author: Love this idea. Thanks to you both!

```ts
import { removeTypenameFromVariables } from '@apollo/client/link/remove-typename';

> const removeTypenameLink = removeTypenameFromVariables();

Reviewer 3: Nit, but maybe something we would still consider for consistency: we have links `HttpLink`, `BatchHttpLink`, `RetryLink` etc., so I was expecting this named export to also be capitalized.

Author: Its funny because you looked at those and I looked at our other links:

- [`onError`](https://www.apollographql.com/docs/react/api/link/apollo-link-error/)
- [`createPersistedQueryLink`](https://www.apollographql.com/docs/react/api/link/persisted-queries/)
- [`setContext`](https://www.apollographql.com/docs/react/api/link/apollo-link-context/)

Unfortunately it looks like we have no established convention. This link really isn't "stateful" so I liked the function approach here and the "verb-y" sound to it (like `setContext` is). Thoughts?

Reviewer 3: Yeah, that's a tricky one. `onError` may have opened the door to different styles here (`setContext` also sort of makes sense but I still find the different naming + casing styles distracting re: `createXyzLink` vs `XyzLink`). Would be great to unify this in v4 but not a blocker here :)

Author: Agreed. Let's unify this in v4 and establish a pattern. Thanks for bringing it up!

```

As an example, let's take the following query. Apollo Client will automatically add `__typename` fields to the query for each field selection set.

```ts
const query = gql`
  query DashboardQuery($id: ID!) {
    dashboard(id: $id) {
      id
      name
    }
  }
`;

const { data } = await client.query({ query, variables: { id: 1 }});
// {
//   "dashboard": {
//     "__typename": "Dashboard",
//     "id": 1,
//     "name": "My Dashboard"
//   }
// }
```

Now let's update this dashboard by sending a mutation to our server. We'll use the `dashboard` as input to our mutation.

```ts
const mutation = gql`
  mutation UpdateDashboardMutation($dashboard: DashboardInput!) {
    updateDashboard(dashboard: $dashboard) {
      id
      name
    }
  }
`;

await client.mutate({ 
  mutation,
  variables: { 
    dashboard: { ...data.dashboard, name: 'My Updated Dashboard' } 
  }
});
```

Without the use of the `removeTypenameFromVariables` link, the server will return an error because `data.dashboard` still contains the `__typename` field.

## Keep `__typename` in JSON scalars

> The [GraphQL spec](https://spec.graphql.org/October2021/#sec-Input-Objects.Type-Validation) disallows input fields that begin with the characters "\_\_" (two underscores). When the input field is a JSON scalar (i.e. a scalar type that accepts raw JSON as input), this restriction is not applied. If your GraphQL server relies on the `__typename` field within a JSON scalar, you can configure the `removeTypenameFromVariables` link to keep `__typename`.

Reviewer 2: ```suggestion
While the [GraphQL spec](https://spec.graphql.org/October2021/#sec-Input-Objects.Type-Validation) disallows input fields that begin with the characters "\_\_" (two underscores), if the input field is a JSON scalar (i.e. a scalar type that accepts raw JSON as input), this restriction is not applied.  
So you might have cases where you want to remove *most* `__typename`s from variables, but your GraphQL server relies on the `__typename` field within a certain JSON scalar.  
In that case, you can configure the `removeTypenameFromVariables` link to keep `__typename` in certain cases.
```

Author: I like this suggestion! Workshopped it a tad more based on this and ended up here: [ba07aa8](https://github.com/apollographql/apollo-client/pull/11104/commits/ba07aa8460d7f719e5a1318ebfa2e99e7c9eff05)

> Note: the JSON scalar type does not need to be literally named `JSON` to be considered a JSON scalar.

Provide an `except` option to map all variable types that should maintain `__typename`. This is done using the `KEEP` sentinel. Each key in the `except` option should correspond to an input type in your GraphQL schema.

> Here we tell the `removeTypenameFromVariables` link to keep all `__typename` fields for any variable that defines itself as a `JSON` scalar type.

Reviewer 2: ```suggestion
Here we tell the `removeTypenameFromVariables` link to keep all `__typename` fields for any variable that is declared as `JSON` type.
```
I'd remove the "scalar" here since this could be any type, not just a scalar. We could also just have an input type named "JSON".

Author: Very true! Good call.

```ts
import { removeTypenameFromVariables, KEEP } from '@apollo/client/link/remove-typename';

const removeTypenameLink = removeTypenameFromVariables({
  except: {
    JSON: KEEP
  }
});
```

> Variable types are inferred from the GraphQL query. Let's take the following GraphQL mutation:

Reviewer 2: General note: before this statement, the reader could also assume that the above `JSON: KEEP` means "keep everything with `__typename: "JSON"`, so maybe this sentence should be a bit further up?

Author: Thanks! This prompted me to reorganize the code blocks too which felt out of order: [e97a386](https://github.com/apollographql/apollo-client/pull/11104/commits/e97a3865f8751f6079594ca220f4e2a8e3e2b2f8)
